:seedling: Redefine managing IAM resources: Create and Delete

Atharva-Shinde opened the pull request:
<!-- Thanks for this PR! If this is your first PR please read the [contributing guide](../CONTRIBUTING.md) -->
<!-- If this PR is still work-in-progress and is being open for visibility please prefix the title with `WIP:` -->

**What type of PR is this?**
/kind feature
/area clusterawsadm

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
/kind regression
/kind support
-->

**What this PR does / why we need it**:
Currently, CAPA manages prerequisites required by AWS through CloudFormation which has caused numerous issues to CAPA end-users. This PR works as a stepping stone in migrating away from the use of AWS CloudFormation and relying on service specific API calls to manage IAM resources and gradually make the process idempotent.

This PR introduces 2 new commands and to enable them set the `CLUSTERAWSADM_ALPHA` environment variable to `true`.

1. `clusterawsadm bootstrap iam create`: creates IAM resources(roles, instances profiles and policies) from the bootstrap configuration file (uses default bootstrap configuration if not provided)
2. `clusterawsadm bootstrap iam delete` deletes IAM resources(roles, instances profiles and policies) created using the bootstrap configuration file (uses default bootstrap configuration if not provided)

<!-- Enter a description of the change and why this change is needed -->

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3715

**Special notes for your reviewer**:
Screenshots:

<img width="1365" alt="Screenshot 2024-09-02 at 9 35 26 PM" src="https://github.com/user-attachments/assets/b7c0e5ec-6764-49fb-b92f-77568fa16e3f">

<img width="1340" alt="Screenshot 2024-09-02 at 9 37 03 PM" src="https://github.com/user-attachments/assets/37fa4386-1f6f-4197-ae48-6bc51b28783c">


**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
3. If no release note is required, just write "NONE".
-->
```release-note

```

Atharva-Shinde commented:
/assign @richardcase @Ankitasw

nrb commented:
/retitle :seedling: Redefine managing IAM resources: Create and Delete

Updating the title should cause the PR verify job to pass.

Atharva-Shinde commented:
/retest

Atharva-Shinde commented:
Thanks @nrb I've addressed the CI failures :)

nrb commented on cmd/clusterawsadm/iamservice/service.go:
```diff
@@ -0,0 +1,500 @@
+
+func attachPoliciesToRole(rolename *string, awsManagedPolicies []string, client *iam.IAM) error {
+	if awsManagedPolicies == nil {
+		// klog.Warningf("no policies defined to attach to the IAM role \"%s\"", *rolename) // TODO
```
What's the next action here? Just enabling the warning, or something else?

nrb replied:
Reading further, looks like there's a question about outputting the ARNs elsewhere - is this the same?

Atharva-Shinde replied:
I think at that time I was wondering if we should issue a warning that the user hasn't defined any awsManagedPolicies in their CloudFormation config or just return back to parent call directly without any warnings.

Atharva-Shinde replied:
So should we give the warning?

nrb commented on cmd/clusterawsadm/iamservice/service.go:
```diff
@@ -0,0 +1,500 @@
+					klog.Warningf("IAM role \"%s\" is already attached to policy", *rolename) // TODO should we output the policy arn? how safe is it
+					continue
+				default:
+					return errors.Wrapf(err, "failed to attach IAM role \"%s\" to policy", *rolename) // TODO should we output the policy arn? how safe is it
```
Thinking about where this command should be running, I can think of a couple environments:

First, a user's machine, while they set up a management cluster. I think outputting an ARN here is fine. They have some sort credentials on their machine, which are the higher risk asset.


Next, CI, which might be spinning up resources, either for testing or as part of a larger provisioning system. In this case, the ARNs would probably get output into logs. Hopefully, the CI is using short-lived credentials, but again, those would be the more valuable asset.

It's true that policy ARNs could be observed and an attacker could attempt to edit the policy to expand permissions. However, I'm not sure the actual ARN is all that important, since what really matters are the permissions to actually overwrite the policy the ARN points to.

Does that make sense?

Atharva-Shinde replied:
Yeah that make sense :) I'll make the change to include the ARNs in the outputs above.

nrb commented on cmd/clusterawsadm/iamservice/service.go:
```diff
@@ -0,0 +1,500 @@
+func prioritySet(t go_cfn.Template) (rmap map[string][]go_cfn.Resource, err error) {
+	rmap = map[string][]go_cfn.Resource{}
+	for _, resource := range t.Resources {
+		if resource.AWSCloudFormationType() == configservice.ResourceTypeAwsIamRole {
```
The description says that this PR is looking to move away from CloudFormation, but we're using CloudFormation files here.

Is the intent to use CloudFormation templates as the input still, but introspect them and make our own API calls vs submitting them directly to the CF service endpoints?

Atharva-Shinde replied:
The intent as discussed with my mentors Richard and Ankita was to think of a solution that eventually gets rid of the CloudFormation to manage IAM resources. The end goal is to create an idempotent solution that enforces API calls to AWS to manage the IAM resources and doesn't rely on CloudFormation Stack. And the intent is to do this incrementally and not all at once so that we avoid big code change altogether and maintain a low to no user side changes.

So this PR is the first step in achieving the above, where we rely on the CF config(template) to get the input and make API calls to manage these IAM resources without relying on CloudFormation stack.

nrb reviewed (commented):
This seems reasonable to me overall. I've added some questions to help me understand where this might go in the future before I approve it.

Atharva-Shinde commented:
/remove-lifecycle stale

richardcase commented:
Sorry @Atharva-Shinde i've not had a huge amount of time recently. I will make sure i spent time reviewing.

nrb commented:
/remove-lifecycle stale

richardcase commented:
/lifecycle hold

We really need this. To get some extra eyes on it:

/help

richardcase commented:
/lifecycle freeze

richardcase commented:
/remove-lifecycle rotten

richardcase commented:
/priority important-soon
/milestone v2.9

richardcase commented:
Myself & @Atharva-Shinde just had a good chat on this. Notes from the call:

Initial items to work on:

1. Move the new command so its marked as alpha

Option 1

this
`clusterawsadm bootstrap iam create`
becomes:
`clusterawsadm alpha bootstrap iam create`

Option 2

```shell
export CLUSTERAWSADM_ALPHA=true
clusterawsadm bootstrap iam create
```


2. Add tests to determine if it created/deleted the IAM resources as expected
3. Ensure updates can support adding a new permission to existing policy


For the future:

- Always add a tag to say "owned" by CAPA

For Rich

- Create a graduation acceptance test to compare resources created by both sets of commands

Atharva-Shinde commented:
I will prioritize this next week.

Atharva-Shinde committed:
feat: CAPA IAM resource management using AWS SDK

Create and delete AWS Identity and Access Management (IAM) resources,
namely roles, instance profiles and policies using AWS SDK.

This feature is part of the clusterawsadm command and is
gated by the environment variable CLUSTERAWSADM_ALPHA
which when set to true is available to use.

export CLUSTERAWSADM_ALPHA=true

clusterawsadm bootstrap iam create

clusterawsadm bootstrap iam delete

Signed-off-by: Atharva Shinde <atharvashinde179@gmail.com>

Atharva-Shinde committed:
test: Add mock and unit tests for creation of IAM resources

Signed-off-by: Atharva Shinde <atharvashinde179@gmail.com>
